novawallet/Modules/Staking/StakingSetupProxy/Validation/ProxyDataValidatorFactory.swift

```diff
@@ -0,0 +1,163 @@
+import Foundation
+import BigInt
+import SoraFoundation
+
+protocol ProxyDataValidatorFactoryProtocol: BaseDataValidatingFactoryProtocol {
+    func hasSufficientBalance(
+        available: BigUInt?,
+        deposit: BigUInt?,
+        fee: BigUInt?,
+        asset: AssetBalanceDisplayInfo,
+        locale: Locale
+    ) -> DataValidating
+
+    func validAddress(
+        _ address: String,
+        chain: ChainModel,
+        locale: Locale
+    ) -> DataValidating
+
+    func notReachedMaximimProxyCount(
+        _ proxyCount: Int?,
+        limit: Int?,
+        chain: ChainModel,
+        locale: Locale
+    ) -> DataValidating
+
+    func proxyNotExists(
+        address: String,
+        chain: ChainModel,
+        proxyList: [Proxy.ProxyDefinition]?,
+        locale: Locale
+    ) -> DataValidating
+}
+
+final class ProxyDataValidatorFactory: ProxyDataValidatorFactoryProtocol {
+    weak var view: ControllerBackedProtocol?
+
+    var basePresentable: BaseErrorPresentable { presentable }
+
+    let presentable: ProxyErrorPresentable
+    let balanceViewModelFactoryFacade: BalanceViewModelFactoryFacadeProtocol
+
+    init(
+        presentable: ProxyErrorPresentable,
+        balanceViewModelFactoryFacade: BalanceViewModelFactoryFacadeProtocol
+    ) {
+        self.presentable = presentable
+        self.balanceViewModelFactoryFacade = balanceViewModelFactoryFacade
+    }
+
+    func hasSufficientBalance(
+        available: BigUInt?,
+        deposit: BigUInt?,
+        fee: BigUInt?,
+        asset: AssetBalanceDisplayInfo,
+        locale: Locale
+    ) -> DataValidating {
+        ErrorConditionViolation(onError: { [weak self] in
+            guard
+                let view = self?.view,
+                let viewModelFactory = self?.balanceViewModelFactoryFacade else {
+                return
+            }
+            let balanceDecimal = available?.decimal(assetInfo: asset)
+            let depositDecimal = deposit?.decimal(assetInfo: asset)
+
+            let balanceModel = viewModelFactory.amountFromValue(
+                targetAssetInfo: asset,
+                value: balanceDecimal ?? 0
+            ).value(for: locale)
+            let depositModel = viewModelFactory.amountFromValue(
+                targetAssetInfo: asset,
+                value: depositDecimal ?? 0
+            ).value(for: locale)
+
+            self?.presentable.presentNotEnoughBalanceForDeposit(
+                from: view,
+                deposit: depositModel,
+                balance: balanceModel,
+                locale: locale
+            )
+        }, preservesCondition: {
+            guard let deposit = deposit,
+                  let fee = fee,
+                  let available = available else {
```

Comment:
I think it is better to require ```available``` to be not nil because we can have cases where balance doesn't exist